### PR TITLE
🏗🚀✨ Enable fast compilation and full sourcemaps for single-pass

### PR DESCRIPTION
**Pull request by rsimha**
**Background:**
- #21939 made speed improvements for multi-pass `gulp dist` by using `nailgun`. However, single-pass `gulp dist` uses the native plugin of `google-closure-compiler` for `v0.js` + `extensions/`, while it concurrently uses the `gulp` plugin for a dozen other compilation targets. This causes problems when  compilation speed up is attempted using `nailgun`.
- #21951 added the `--full_sourcemaps` flag to `gulp dist` to enable the generation of sourcemaps with full source code. However, single-pass neither generates sourcemaps, nor does it use the `--full_sourcemaps` mechanism.

**Changes in this PR:**
- Refactors / cleans up all compilation code into `build-system/compile/`
- Makes single-pass compilation use the same closure compiler plugin as multi-pass
- Enables the `nailgun` speed up
- Enables the `--full_sourcemaps` flag

**Running time improvements:**

- **Macbook Pro:** [~2.2 minutes](https://gist.github.com/rsimha/fb9175ece4588d939aeb823f122d7f4c) → [~1.8 minutes](https://gist.github.com/rsimha/ce9800ecb07e6a292a88ac5496638632)
- **Travis CI:** [~4 minutes](https://travis-ci.org/ampproject/amphtml/jobs/525177861#L356) → [~2.6 minutes](https://travis-ci.org/ampproject/amphtml/jobs/525546081#L368)

This results in a ~1.5x speed up for `gulp dist --single_pass`.

Follow up to #21939
Follow up to #21951

### CODE REVIEW

**Comment by rsimha** [comment]
Update: It looks like the single-pass build fails when part of the compilation is done with `nailgun` and the rest of it with `runner.jar`. It should be possible to do all compilation with `nailgun`. Coming up.

https://travis-ci.org/ampproject/amphtml/jobs/524584657#L362

**Comment by rsimha** [comment]
Switched single pass compilation to use `nailgun` across all compiler invocations, but I'm still seeing problems that I believe are due to a race between the `compile()` in `compile.js` and the one in `single-pass.js`. Still debugging.

https://travis-ci.org/ampproject/amphtml/jobs/524803833#L362

**Comment by kristoferbaxter** [comment]
I'd the native version slower than using nailgun? I'd be highly surprised, since they both share the goal of lowering the startup cost for the VM.

In fact, my hunch is if everything used the native version it be faster. Can we test this before adding another dependency?

**Comment by rsimha** [comment]
@kristoferbaxter `gulp dist --single_pass` with `nailgun` is ~1.5x faster than with the native `runner.jar`. This is because single pass does one giant invocation for `v0.js` + `extensions/`, and about a dozen other invocations for these targets:
```js
[03:46:46] Minified recaptcha.js (12.808 s)
[03:46:46] Minified iframe-transport-client-v0.js (12.835 s)
[03:46:46] Minified ampcontext-v0.js (12.851 s)
[03:46:52] Minified f.js (18.955 s)
[03:47:22] Minified video-iframe-integration-v0.js (48.371 s)
[03:47:22] Minified amp4ads-host-v0.js (49.015 s)
[03:47:23] Minified alp.js (49.693 s)
[03:47:28] Minified examiner.js (54.767 s)
[03:47:30] Minified ww.js (56.376 s)
[03:47:31] Minified experiments.js (56.401 s)
[03:47:34] Minified amp-web-push-helper-frame.js (58.703 s)
[03:47:35] Minified amp-web-push-permission-dialog.js (59.786 s)
[03:47:37] Minified amp-login-done-0.1.js → amp-login-done-latest.js (804 ms)
```

**Comment by rsimha** [comment]
This is now ready for review.

File: build-system/compile/single-pass.js [hunk 1]
```diff
@@ -142,17 +143,20 @@ exports.getFlags = function(config) {
   return exports.getGraph(config.modules, config).then(function(g) {
     return flagsArray.concat(
         exports.getBundleFlags(g, flagsArray));
+  }).then(flagsArray => {
+    fs.writeFileSync('flags-array.txt', JSON.stringify(flagsArray, null, 2));
```
Comment (cramforce) [inline build-system/compile/single-pass.js]: I wonder if we should just remove this. Seems to be basically left over debugging code.

Comment (rsimha) [reply build-system/compile/single-pass.js]: Done.

**Comment by kristoferbaxter** [comment]
We can discuss more in person, but have you evaluated running closure compiler natively (not runner jar) similar to how the Rollup plugin we publish does it?

This avoids JVM start-up costs, and doesn't introduce nailgun to the project.

I'm back in the office next week and will follow-up with you in person.

**Comment by rsimha** [comment]
@kristoferbaxter I looked at the code in `ampproject/rollup-plugin-closure-compiler`, particularly [this code](https://github.com/ampproject/rollup-plugin-closure-compiler/blob/e979a438fe11914992fb8951017cdc5c9e2c4c97/src/compiler.ts#L49-L52) in `compiler.ts`.

- This won't work for `amphtml` because the `runner.jar` we use does a whole bunch of custom transforms that are missing from the default `compiler.jar` in `google-closure-compiler`.
- From reading the code in `google-closure-compiler`, I'm not convinced that using `rollup-plugin-closure-compiler` across multiple concurrent compilation invocations will actually reduce the JVM overhead because each invocation internally [spawns](https://github.com/google/closure-compiler-npm/blob/v20190301.0.0/packages/google-closure-compiler/lib/node/closure-compiler.js#L76) a new java process. (We make ~15 compiler invocations during `gulp dist --single_pass` and a few dozen invocations during `gulp dist`.)

Happy to chat next week.

**Comment by cramforce** [comment]
There is a natively compiled closure compiler which we could in theory provide, but

- it is unlikely to be faster than a resident JVM (We benefit a lot from the JIT getting really good at running closure compiler)
- has issues with cross-platform compatibility

**Comment by rsimha** [comment]
@kristoferbaxter If you want to compare compile times with and without `nailgun`, you can now do so using the standalone `gulp` task. See https://github.com/ampproject/amphtml/pull/21977#issue-273385192

**Comment by kristoferbaxter** [comment]
The nailgun changes are interesting, but we really want out of custom transforms done using closure compiler. 

In the single pass build, they are not allowed... Instead all transforms should be accomplished using more well known tools (Babel).

Additionally, each process spawned is not spinning up a JVM when using the native executable.

For single-pass, our goal is to remove as much custom configuration and work to accomplish a build. This change isn't in that spirit, and we should ensure it's enough of an impact versus the originally agreed path (native CC dist and check types is a separate job).

Happy to continue the conversation this week in person.

**Comment by rsimha** [comment]
@kristoferbaxter I see. That was useful. Here's my understanding of what you're looking for:

- `gulp dist --single_pass` invokes `runner.jar` (now wrapped in `nailgun`) ~15 times.
- For [14 of the 15 invocations](https://github.com/ampproject/amphtml/pull/21953#issuecomment-487355208), we must use `runner.jar` because of the [custom transforms](https://github.com/ampproject/amphtml/blob/680dd188e47114af397083500333a7ab0ddd886e/build-system/runner/src/org/ampproject/AmpCodingConvention.java#L33-L127) it applies
- One of those invocations is special (the one that minifies `v0.js` + `extensions/`). For it, we'd like to use the native closure compiler binary. (On Mac OS, this is `node_modules/google-closure-compiler-osx/compiler`, which is not a `.jar` file)

I tried this, and it's currently broken because of a closure compiler bug (cannot find reference types by relative path).
- For context, see  https://github.com/google/closure-compiler/issues/3041, https://github.com/google/closure-compiler/commit/acea045bcc566c8d1ba74bc8f63fbb1568a18c2b, and https://github.com/ampproject/amphtml/pull/21618#issue-265230068.
- For the 2000+ errors, see the [build output](https://gist.github.com/rsimha/5d0114fe0d6b99efcbab6133eee02bd1).
- I saw the [same errors](https://gist.github.com/rsimha/4f9e105936f3eac4bba6b13cbbe3ec48) when I used the default java `compiler.jar` instead of the native compiler

Once the bug is fixed, I'm open to moving away from `runner.jar` for the one special invocation (we'd still have to use it for the other 14).

Regarding the impact of this change, most of the speed up comes from the 14 invocations of `runner.jar`. On Travis CI, running time was reduced from [~4 minutes](https://travis-ci.org/ampproject/amphtml/jobs/525177861#L356) to [~2.6 minutes](https://travis-ci.org/ampproject/amphtml/jobs/525546081#L368), a ~1.5x speed up.

**Pull request merged**